x-pack/plugins/task_manager/server/monitoring/capacity_estimation.ts

```diff
@@ -0,0 +1,245 @@
+      (assumedKibanaInstances / minRequiredKibanaInstances) +
+    averageRecurringRequiredPerMinute / minRequiredKibanaInstances;
+
+  const assumedAverageRecurringRequiredThroughputPerMinutePerKibana =
+    averageRecurringRequiredPerMinute / assumedKibanaInstances;
+  /**
+   * assuming the historical capacity needed for ephemeral and non-recurring tasks, plus
+   * the amount we know each kibana would need for recurring tasks, how much capacity would
+   * each kibana need if the assumed current number were correct?
+   */
+  const assumedRequiredThroughputPerMinutePerKibana =
+    averageCapacityUsedByNonRecurringAndEphemeralTasksPerKibana +
+    averageRecurringRequiredPerMinute / assumedKibanaInstances;
+
+  return {
+    status:
+      assumedRequiredThroughputPerMinutePerKibana < capacityPerMinutePerKibana
+        ? HealthStatus.OK
+        : assumedAverageRecurringRequiredThroughputPerMinutePerKibana < capacityPerMinutePerKibana
+        ? HealthStatus.Warning
+        : HealthStatus.Error,
+    timestamp: new Date().toISOString(),
+    value: {
+      observed: mapValues(
+        {
+          observed_kibana_instances: assumedKibanaInstances,
+          max_throughput_per_minute_per_kibana: capacityPerMinutePerKibana,
+          max_throughput_per_minute: assumedCapacityAvailablePerMinute,
+          minutes_to_drain_overdue:
+            overdue / (assumedKibanaInstances * averageCapacityUsedByPersistedTasksPerKibana),
+          avg_recurring_required_throughput_per_minute: averageRecurringRequiredPerMinute,
+          avg_recurring_required_throughput_per_minute_per_kibana: assumedAverageRecurringRequiredThroughputPerMinutePerKibana,
+          avg_required_throughput_per_minute:
+            assumedRequiredThroughputPerMinutePerKibana * assumedKibanaInstances,
+          avg_required_throughput_per_minute_per_kibana: assumedRequiredThroughputPerMinutePerKibana,
+        },
+        Math.ceil
+      ),
+      proposed: mapValues(
+        {
+          proposed_kibana: minRequiredKibanaInstances,
+          min_required_kibana: minRequiredKibanaInstancesForRecurringTasks,
+          avg_recurring_required_throughput_per_minute_per_kibana: averageRecurringRequiredPerMinutePerKibana,
+          avg_required_throughput_per_minute_per_kibana: averageRequiredThroughputPerMinutePerKibana,
+        },
+        Math.ceil
+      ),
+    },
+  };
+}
+
+export function withCapacityEstimate(
+  monitoredStats: RawMonitoringStats['stats']
+): RawMonitoringStats['stats'] {
+  if (isCapacityEstimationParams(monitoredStats)) {
+    return {
+      ...monitoredStats,
+      capacity_estimation: estimateCapacity(monitoredStats),
+    };
+  }
+  return monitoredStats;
+}
+
+function percentageOf(val: number, percentage: number) {
+  return Math.round((percentage * val) / 100);
+}
+
+function getAverageDuration(durations: Partial<TaskPersistenceTypes<AveragedStat>>): number {
+  return stats.mean(
```

Comment:
I noticed `stats.mean([])` => `NaN`, which is why you have the `isFinite()` test on the result where this function is used above.  Wondering if we should have it return something else, in case this function gets used later by something else, but the code doesn't check for the `NaN`.  Return `0` seems ... not great.  Maybe `undefined`, which we could add to the return type signature so you'd be forced to check.

Also, I'm not familiar with `isFinite()`, guessing other folk may not be either, so might be good to change for that reason.  And actually, I'm curious, why not use `isNaN()` instead of `isFinite()` - is there a possibility this number could be `Infinity`?

Reply:
> is there a possibility this number could be Infinity?

Yup :)
I found some edge cases, such as when the % of recurring tasks ran in the running average dropped to zero and I divided by 0.

I was looking for an elegant way... I can wrap it with a Result.. that's the Rust way ;)

Reply:
Looked closer, in this specific case no chance of an Infinity I don't think, but there was another use I had made that needed it.
I've now refactored and no more need for isFinite. :)